File: backend/models/calendar.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, timedelta
from enum import Enum
from .base import BaseDocument, PyObjectId

class EventType(str, Enum):
    TASK_DEADLINE = "task_deadline"
    PROJECT_MILESTONE = "project_milestone"
    MEETING = "meeting"
    RECORDING_SESSION = "recording_session"
    PERFORMANCE = "performance"
    REHEARSAL = "rehearsal"
    MARKETING_CAMPAIGN = "marketing_campaign"
    RELEASE_DATE = "release_date"
    TOUR_DATE = "tour_date"
    PERSONAL = "personal"
    OTHER = "other"

class EventStatus(str, Enum):
    SCHEDULED = "scheduled"
    CONFIRMED = "confirmed"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    POSTPONED = "postponed"

class EventPriority(str, Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"

class RecurrencePattern(BaseModel):
    frequency: str  # daily, weekly, monthly, yearly
    interval: int = 1  # every N frequency units
    days_of_week: Optional[List[int]] = None  # 0=Monday, 6=Sunday
    day_of_month: Optional[int] = None
    end_date: Optional[datetime] = None
    occurrence_count: Optional[int] = None

class EventReminder(BaseModel):
    minutes_before: int
    method: str = "notification"  # notification, email, sms
    sent: bool = False

class EventAttendee(BaseModel):
    user_id: Optional[PyObjectId] = None
    email: Optional[str] = None
    name: str
    status: str = "pending"  # pending, accepted, declined, tentative
    role: Optional[str] = None

class EventLocation(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = None
    venue_type: Optional[str] = None
    coordinates: Optional[Dict[str, float]] = None  # lat, lng

class EventBase(BaseModel):
    title: str
    description: Optional[str] = None
    type: EventType
    status: EventStatus = EventStatus.SCHEDULED
    priority: EventPriority = EventPriority.MEDIUM
    start_time: datetime
    end_time: datetime
    all_day: bool = False
    location: Optional[EventLocation] = None
    tags: List[str] = []

class EventCreate(EventBase):
    project_id: Optional[PyObjectId] = None
    task_id: Optional[PyObjectId] = None
    attendees: List[EventAttendee] = []
    reminders: List[EventReminder] = []
    recurrence: Optional[RecurrencePattern] = None

class EventUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[EventStatus] = None
    priority: Optional[EventPriority] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    all_day: Optional[bool] = None
    location: Optional[EventLocation] = None
    tags: Optional[List[str]] = None

class EventInDB(BaseDocument, EventBase):
    creator_id: PyObjectId
    project_id: Optional[PyObjectId] = None
    task_id: Optional[PyObjectId] = None
    attendees: List[EventAttendee] = []
    reminders: List[EventReminder] = []
    recurrence: Optional[RecurrencePattern] = None
    parent_event_id: Optional[PyObjectId] = None  # For recurring events
    is_recurring: bool = False
    external_calendar_id: Optional[str] = None
    external_event_id: Optional[str] = None
    ai_generated: bool = False
    conflict_detected: bool = False
    travel_time_minutes: Optional[int] = None

class Event(EventBase):
    id: PyObjectId = Field(alias="_id")
    creator_id: PyObjectId
    project_id: Optional[PyObjectId] = None
    task_id: Optional[PyObjectId] = None
    attendees: List[EventAttendee] = []
    reminders: List[EventReminder] = []
    recurrence: Optional[RecurrencePattern] = None
    parent_event_id: Optional[PyObjectId] = None
    is_recurring: bool = False
    external_calendar_id: Optional[str] = None
    external_event_id: Optional[str] = None
    ai_generated: bool = False
    conflict_detected: bool = False
    travel_time_minutes: Optional[int] = None
    created_at: datetime
    updated_at: datetime

# Calendar Integration Models
class CalendarIntegrationBase(BaseModel):
    provider: str  # google, outlook, apple, etc.
    calendar_id: str
    calendar_name: str
    is_primary: bool = False
    sync_enabled: bool = True
    sync_direction: str = "bidirectional"  # import_only, export_only, bidirectional

class CalendarIntegrationCreate(CalendarIntegrationBase):
    access_token: str
    refresh_token: Optional[str] = None

class CalendarIntegrationUpdate(BaseModel):
    calendar_name: Optional[str] = None
    is_primary: Optional[bool] = None
    sync_enabled: Optional[bool] = None
    sync_direction: Optional[str] = None

class CalendarIntegrationInDB(BaseDocument, CalendarIntegrationBase):
    user_id: PyObjectId
    access_token: str
    refresh_token: Optional[str] = None
    token_expires_at: Optional[datetime] = None
    last_sync: Optional[datetime] = None
    sync_status: str = "active"  # active, error, disabled
    sync_error: Optional[str] = None

class CalendarIntegration(CalendarIntegrationBase):
    id: PyObjectId = Field(alias="_id")
    user_id: PyObjectId
    token_expires_at: Optional[datetime] = None
    last_sync: Optional[datetime] = None
    sync_status: str = "active"
    sync_error: Optional[str] = None
    created_at: datetime
    updated_at: datetime

File: types/calendar.ts
export interface CalendarEvent {
  id: string;
  title: string;
  description?: string;
  type: EventType;
  status: EventStatus;
  priority: EventPriority;
  start_time: string;
  end_time: string;
  all_day: boolean;
  location?: EventLocation;
  tags: string[];
  creator_id: string;
  project_id?: string;
  task_id?: string;
  attendees: EventAttendee[];
  reminders: EventReminder[];
  recurrence?: RecurrencePattern;
  parent_event_id?: string;
  is_recurring: boolean;
  external_calendar_id?: string;
  external_event_id?: string;
  ai_generated: boolean;
  conflict_detected: boolean;
  travel_time_minutes?: number;
  created_at: string;
  updated_at: string;
}

export type EventType = 
  | 'task_deadline'
  | 'project_milestone'
  | 'meeting'
  | 'recording_session'
  | 'performance'
  | 'rehearsal'
  | 'marketing_campaign'
  | 'release_date'
  | 'tour_date'
  | 'personal'
  | 'other';

export type EventStatus = 
  | 'scheduled'
  | 'confirmed'
  | 'in_progress'
  | 'completed'
  | 'cancelled'
  | 'postponed';

export type EventPriority = 
  | 'critical'
  | 'high'
  | 'medium'
  | 'low';

export interface EventLocation {
  name?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  venue_type?: string;
  coordinates?: {
    lat: number;
    lng: number;
  };
}

export interface EventAttendee {
  user_id?: string;
  email?: string;
  name: string;
  status: 'pending' | 'accepted' | 'declined' | 'tentative';
  role?: string;
}

export interface EventReminder {
  minutes_before: number;
  method: 'notification' | 'email' | 'sms';
  sent: boolean;
}

export interface RecurrencePattern {
  frequency: 'daily' | 'weekly' | 'monthly' | 'yearly';
  interval: number;
  days_of_week?: number[];
  day_of_month?: number;
  end_date?: string;
  occurrence_count?: number;
}

export interface CreateEventData {
  title: string;
  description?: string;
  type: EventType;
  status?: EventStatus;
  priority?: EventPriority;
  start_time: string;
  end_time: string;
  all_day?: boolean;
  location?: EventLocation;
  tags?: string[];
  project_id?: string;
  task_id?: string;
  attendees?: EventAttendee[];
  reminders?: EventReminder[];
  recurrence?: RecurrencePattern;
}

export interface UpdateEventData {
  title?: string;
  description?: string;
  status?: EventStatus;
  priority?: EventPriority;
  start_time?: string;
  end_time?: string;
  all_day?: boolean;
  location?: EventLocation;
  tags?: string[];
}

export interface CalendarIntegration {
  id: string;
  provider: string;
  calendar_id: string;
  calendar_name: string;
  is_primary: boolean;
  sync_enabled: boolean;
  sync_direction: 'import_only' | 'export_only' | 'bidirectional';
  user_id: string;
  token_expires_at?: string;
  last_sync?: string;
  sync_status: 'active' | 'error' | 'disabled';
  sync_error?: string;
  created_at: string;
  updated_at: string;
}